src/browser/assistant/search.js

```diff
@@ -0,0 +1,308 @@
+/* @flow */
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+
+import {Effects, Task, html, forward, thunk} from "reflex";
+import {merge, always, tag, tagged, batch} from "../../common/prelude";
+import {Style, StyleSheet} from '../../common/style';
+import {indexOfOffset} from "../../common/selector";
+import * as Result from '../../common/result';
+
+import * as Title from "./title";
+import * as Icon from "./icon";
+import * as Suggestion from "./suggestion";
+import * as Unknown from '../../common/unknown';
+
+/*::
+import * as Search from "../../../type/browser/assistant/search"
+*/
+
+
+const Abort = tag("Abort");
+export const SelectNext = tagged("SelectNext");
+export const SelectPrevious = tagged("SelectPrevious");
+export const Suggest = tag("Suggest");
+export const Query = tag("Query");
+export const Execute = tag("Execute");
+export const Activate = tag("Activate");
+const UpdateMatches = tag("UpdateMatches");
+const byURI =
+  uri =>
+  action =>
+  tagged("ByURI", {uri, action});
+
+const decodeFailure = ({target: request}) =>
+  Result.error
+  // @FlowIssue: Flow does not know about `request.url`
+  (Error(`Failed to send request to ${request.url} : ${request.statusText}`));
+
+const decodeResponseFailure =
+  request =>
+  // @FlowIssue: Flow does not know about `request.response`
+  Result.error(Error(`Can not decode ${request.respose} received from ${request.url}`))
+
+const decodeMatches =
+  matches =>
+  ( Array.isArray(matches)
+  ? Result.ok(matches.map(decodeMatch))
+  : Result.error(Error(`Can not decode non array matches ${matches}`))
+  );
+
+const decodeMatch =
+  match =>
+  ( { title: match
+    , uri: `https://duckduckgo.com/?q=${encodeURIComponent(match)}`
+    }
+  );
+
+const decodeResponse = ({target: request}) =>
+  // @FlowIssue: Flow does not know about `request.responseType`
+  ( request.responseType !== 'json'
+  // @FlowIssue: Flow does not know about `request.url`
+  ? Result.error(Error(`Can not decode ${request.responseType} type response from ${request.url}`))
+  : request.response == null
+  ? decodeResponseFailure(request)
+  : request.response[1] == null
+  ? decodeResponseFailure(request)
+  : decodeMatches(request.response[1])
+  );
+
+const pendingRequests = Object.create(null);
+
+const abort =
+  id =>
+  Task.io(deliver => {
+    if (pendingRequests[id] != null) {
+      pendingRequests[id].abort();
+      delete pendingRequests[id];
+    }
+  })
+
+const search/*:Search.search*/ =
+  (id, input, limit) =>
+  Task.future(() => new Promise(resolve => {
+    const request = new XMLHttpRequest({ mozSystem: true });
```

Comment:
Nit: we should have a comment that explains we use XHR instead of `fetch` because we need to set `mozSystem: true`.